File: src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxTreeTests.cs
// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.

using System.Text;
using Microsoft.CodeAnalysis.Text;
using Xunit;
using Roslyn.Test.Utilities;
using Roslyn.Utilities;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using System.Collections.Generic;
using System;

namespace Microsoft.CodeAnalysis.CSharp.UnitTests
{
    public class SyntaxTreeTests
    {
        [Fact]
        public void WithRootAndOptions_ParsedTree()
        {
            var oldTree = SyntaxFactory.ParseSyntaxTree("class B {}");
            var newRoot = SyntaxFactory.ParseCompilationUnit("class C {}");
            var newOptions = new CSharpParseOptions();
            var newTree = oldTree.WithRootAndOptions(newRoot, newOptions);
            var newText = newTree.GetText();

            Assert.Equal(newRoot.ToString(), newTree.GetRoot().ToString());
            Assert.Same(newOptions, newTree.Options);

            Assert.Null(newText.Encoding);
            Assert.Equal(SourceHashAlgorithm.Sha1, newText.ChecksumAlgorithm);
        }

        [Fact]
        public void WithRootAndOptions_ParsedTreeWithText()
        {
            var oldText = SourceText.From("class B {}", Encoding.UTF7, SourceHashAlgorithm.Sha256);
            var oldTree = SyntaxFactory.ParseSyntaxTree(oldText);

            var newRoot = SyntaxFactory.ParseCompilationUnit("class C {}");
            var newOptions = new CSharpParseOptions();
            var newTree = oldTree.WithRootAndOptions(newRoot, newOptions);
            var newText = newTree.GetText();

            Assert.Equal(newRoot.ToString(), newTree.GetRoot().ToString());
            Assert.Same(newOptions, newTree.Options);
            Assert.Same(Encoding.UTF7, newText.Encoding);
            Assert.Equal(SourceHashAlgorithm.Sha256, newText.ChecksumAlgorithm);
        }

        [Fact]
        public void WithRootAndOptions_DummyTree()
        {
            var dummy = new CSharpSyntaxTree.DummySyntaxTree();
            var newRoot = SyntaxFactory.ParseCompilationUnit("class C {}");
            var newOptions = new CSharpParseOptions();
            var newTree = dummy.WithRootAndOptions(newRoot, newOptions);
            Assert.Equal(newRoot.ToString(), newTree.GetRoot().ToString());
            Assert.Same(newOptions, newTree.Options);
        }

        [Fact]
        public void WithFilePath_ParsedTree()
        {
            var oldTree = SyntaxFactory.ParseSyntaxTree("class B {}", path: "old.cs");
            var newTree = oldTree.WithFilePath("new.cs");
            var newText = newTree.GetText();

            Assert.Equal(newTree.FilePath, "new.cs");
            Assert.Equal(oldTree.ToString(), newTree.ToString());

            Assert.Null(newText.Encoding);
            Assert.Equal(SourceHashAlgorithm.Sha1, newText.ChecksumAlgorithm);
        }

        [Fact]
        public void WithFilePath_ParsedTreeWithText()
        {
            var oldText = SourceText.From("class B {}", Encoding.UTF7, SourceHashAlgorithm.Sha256);
            var oldTree = SyntaxFactory.ParseSyntaxTree(oldText, path: "old.cs");

            var newTree = oldTree.WithFilePath("new.cs");
            var newText = newTree.GetText();

            Assert.Equal(newTree.FilePath, "new.cs");
            Assert.Equal(oldTree.ToString(), newTree.ToString());

            Assert.Same(Encoding.UTF7, newText.Encoding);
            Assert.Equal(SourceHashAlgorithm.Sha256, newText.ChecksumAlgorithm);
        }

        [Fact]
        public void WithFilePath_DummyTree()
        {
            var oldTree = new CSharpSyntaxTree.DummySyntaxTree();
            var newTree = oldTree.WithFilePath("new.cs");

            Assert.Equal(newTree.FilePath, "new.cs");
            Assert.Equal(oldTree.ToString(), newTree.ToString());
        }

        [Fact, WorkItem(4576, "https://github.com/dotnet/roslyn/issues/4576")]
        public void VisitLongSequenceOfBinaryExpressions()
        {
            var tree = SyntaxFactory.ParseSyntaxTree(
@"
class Test
{ 
    void Main()
    {
" + $"        string s = a0{ BuildSequenceOfBinaryExpressions("a") };" + @"
    }
}
");
            Assert.True(tree.GetDiagnostics().IsEmpty());

            var visitor = new TestWalker();
            visitor.Visit(tree.GetRoot());
            Assert.Equal(10001, visitor.AIdentifiers.Count);

            var internalrewriter = new TestInternalRewriter();
            var result1 = internalrewriter.Visit((Syntax.InternalSyntax.CSharpSyntaxNode)tree.GetRoot().Green);
            Assert.Equal(
@"
class Test
{ 
    void Main()
    {
" + $"        string s = b0{ BuildSequenceOfBinaryExpressions("b") };" + @"
    }
}
", result1.ToFullString());

            var rewriter = new TestRewriter();
            var result2 = rewriter.Visit(tree.GetRoot());

            Assert.Equal(
@"
class Test
{ 
    void Main()
    {
" + $"        string s = c0{ BuildSequenceOfBinaryExpressions("c") };" + @"
    }
}
", result2.ToFullString());
        }

        private static string BuildSequenceOfBinaryExpressions(string identifierPrefix)
        {
            var builder = new StringBuilder();

            for (int i = 1; i <= 10000; i++)
            {
                builder.Append(" + ");
                builder.Append(identifierPrefix);
                builder.Append(i);
            }

            return builder.ToString();
        }

        class TestWalker : CSharpSyntaxWalker
        {
            public HashSet<IdentifierNameSyntax> AIdentifiers = new HashSet<IdentifierNameSyntax>();

            public override void VisitIdentifierName(IdentifierNameSyntax node)
            {
                if (node.Identifier.ValueText.StartsWith("a", StringComparison.Ordinal))
                {
                    AIdentifiers.Add(node);
                }

                base.VisitIdentifierName(node);
            }
        }

        class TestRewriter : CSharpSyntaxRewriter
        {
            public override SyntaxNode VisitIdentifierName(IdentifierNameSyntax node)
            {
                if (node.Identifier.ValueText.StartsWith("a", StringComparison.Ordinal))
                {
                    return SyntaxFactory.IdentifierName(SyntaxFactory.Identifier("c" + node.Identifier.ValueText.Substring(1)).
                                                                          WithLeadingTrivia(node.Identifier.LeadingTrivia).
                                                                          WithTrailingTrivia(node.Identifier.TrailingTrivia));
                }

                return base.VisitIdentifierName(node);
            }
        }

        private class TestInternalRewriter : Syntax.InternalSyntax.CSharpSyntaxRewriter
        {
            public override Syntax.InternalSyntax.CSharpSyntaxNode VisitIdentifierName(Syntax.InternalSyntax.IdentifierNameSyntax node)
            {
                if (node.Identifier.ValueText.StartsWith("a", StringComparison.Ordinal))
                {
                    return Syntax.InternalSyntax.SyntaxFactory.IdentifierName(Syntax.InternalSyntax.SyntaxFactory.Identifier("b" + node.Identifier.ValueText.Substring(1)).
                                                                                      WithLeadingTrivia(node.Identifier.GetLeadingTrivia()).
                                                                                      WithTrailingTrivia(node.Identifier.GetTrailingTrivia()));
                }

                return base.VisitIdentifierName(node);
            }
        }
    }
}

File: src/Compilers/VisualBasic/Test/Syntax/Syntax/SyntaxTreeTests.vb
' Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.

Imports System.Text
Imports Microsoft.CodeAnalysis.Text
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax
Imports Roslyn.Test.Utilities

Namespace Microsoft.CodeAnalysis.VisualBasic.UnitTests
    Public Class VisualBasicSyntaxTreeTests
        <Fact>
        Public Sub WithRootAndOptions_ParsedTree()
            Dim oldTree = SyntaxFactory.ParseSyntaxTree("Class B : End Class")
            Dim newRoot = SyntaxFactory.ParseCompilationUnit("Class C : End Class")
            Dim newOptions = New VisualBasicParseOptions()
            Dim newTree = oldTree.WithRootAndOptions(newRoot, newOptions)
            Dim newText = newTree.GetText()

            Assert.Equal(newRoot.ToString(), newTree.GetRoot().ToString())
            Assert.Same(newOptions, newTree.Options)

            Assert.Null(newText.Encoding)
            Assert.Equal(SourceHashAlgorithm.Sha1, newText.ChecksumAlgorithm)
        End Sub

        <Fact>
        Public Sub WithRootAndOptions_ParsedTreeWithText()
            Dim oldText = SourceText.From("Class B : End Class", Encoding.UTF7, SourceHashAlgorithm.Sha256)
            Dim oldTree = SyntaxFactory.ParseSyntaxTree(oldText)

            Dim newRoot = SyntaxFactory.ParseCompilationUnit("Class C : End Class")
            Dim newOptions = New VisualBasicParseOptions()
            Dim newTree = oldTree.WithRootAndOptions(newRoot, newOptions)
            Dim newText = newTree.GetText()

            Assert.Equal(newRoot.ToString(), newTree.GetRoot().ToString())
            Assert.Same(newOptions, newTree.Options)

            Assert.Same(Encoding.UTF7, newText.Encoding)
            Assert.Equal(SourceHashAlgorithm.Sha256, newText.ChecksumAlgorithm)
        End Sub

        <Fact>
        Public Sub WithRootAndOptions_DummyTree()
            Dim dummy = New VisualBasicSyntaxTree.DummySyntaxTree()
            Dim newRoot = SyntaxFactory.ParseCompilationUnit("Class C : End Class")
            Dim newOptions = New VisualBasicParseOptions()
            Dim newTree = dummy.WithRootAndOptions(newRoot, newOptions)
            Assert.Equal(newRoot.ToString(), newTree.GetRoot().ToString())
            Assert.Same(newOptions, newTree.Options)
        End Sub

        <Fact>
        Public Sub WithFilePath_ParsedTree()
            Dim oldTree = SyntaxFactory.ParseSyntaxTree("Class B : End Class", path:="old.vb")
            Dim newTree = oldTree.WithFilePath("new.vb")
            Dim newText = newTree.GetText()

            Assert.Equal(newTree.FilePath, "new.vb")
            Assert.Equal(oldTree.ToString(), newTree.ToString())

            Assert.Null(newText.Encoding)
            Assert.Equal(SourceHashAlgorithm.Sha1, newText.ChecksumAlgorithm)
        End Sub

        <Fact>
        Public Sub WithFilePath_ParsedTreeWithText()
            Dim oldText = SourceText.From("Class B : End Class", Encoding.UTF7, SourceHashAlgorithm.Sha256)
            Dim oldTree = SyntaxFactory.ParseSyntaxTree(oldText, path:="old.vb")
            Dim newTree = oldTree.WithFilePath("new.vb")
            Dim newText = newTree.GetText()

            Assert.Equal(newTree.FilePath, "new.vb")
            Assert.Equal(oldTree.ToString(), newTree.ToString())

            Assert.Same(Encoding.UTF7, newText.Encoding)
            Assert.Equal(SourceHashAlgorithm.Sha256, newText.ChecksumAlgorithm)
        End Sub

        <Fact>
        Public Sub WithFilePath_DummyTree()
            Dim oldTree = New VisualBasicSyntaxTree.DummySyntaxTree()
            Dim newTree = oldTree.WithFilePath("new.vb")

            Assert.Equal(newTree.FilePath, "new.vb")
            Assert.Equal(oldTree.ToString(), newTree.ToString())
        End Sub

        <Fact, WorkItem(4576, "https://github.com/dotnet/roslyn/issues/4576")>
        Public Sub VisitLongSequenceOfBinaryExpressions()
            Dim tree = SyntaxFactory.ParseSyntaxTree(
$"
Module Test
    Sub Main()
        dim s as string = a0{BuildSequenceOfBinaryExpressions("a")}
    End Sub
End Module
")
            Assert.True(tree.GetDiagnostics().IsEmpty())

            Dim visitor As New TestWalker()
            visitor.Visit(tree.GetRoot())
            Assert.Equal(10001, visitor.AIdentifiers.Count)

            Dim internalrewriter = New TestInternalRewriter()
            Dim result1 = internalrewriter.Visit(DirectCast(tree.GetRoot().Green, InternalSyntax.VisualBasicSyntaxNode))
            Assert.Equal(
$"
Module Test
    Sub Main()
        dim s as string = b0{BuildSequenceOfBinaryExpressions("b")}
    End Sub
End Module
", result1.ToFullString())

            Dim rewriter As New TestRewriter()
            Dim result2 = rewriter.Visit(tree.GetRoot())

            Assert.Equal(
$"
Module Test
    Sub Main()
        dim s as string = c0{BuildSequenceOfBinaryExpressions("c")}
    End Sub
End Module
", result2.ToFullString())
        End Sub

        Private Shared Function BuildSequenceOfBinaryExpressions(identifierPrefix As String) As String
            Dim builder = New StringBuilder()

            For i As Integer = 1 To 10000
                builder.Append(" & ")
                builder.Append(identifierPrefix)
                builder.Append(i)
            Next

            Return builder.ToString()
        End Function

        Class TestWalker
            Inherits VisualBasicSyntaxWalker

            Public AIdentifiers As New HashSet(Of IdentifierNameSyntax)

            Public Overrides Sub VisitIdentifierName(node As IdentifierNameSyntax)
                If node.Identifier.ValueText.StartsWith("a", StringComparison.Ordinal) Then
                    AIdentifiers.Add(node)
                End If

                MyBase.VisitIdentifierName(node)
            End Sub
        End Class

        Class TestRewriter
            Inherits VisualBasicSyntaxRewriter

            Public Overrides Function VisitIdentifierName(node As IdentifierNameSyntax) As SyntaxNode
                If node.Identifier.ValueText.StartsWith("a", StringComparison.Ordinal) Then
                    Return SyntaxFactory.IdentifierName(SyntaxFactory.Identifier("c" + node.Identifier.ValueText.Substring(1)).
                                                                          WithLeadingTrivia(node.Identifier.LeadingTrivia).
                                                                          WithTrailingTrivia(node.Identifier.TrailingTrivia))
                End If

                Return MyBase.VisitIdentifierName(node)
            End Function
        End Class

        Private Class TestInternalRewriter
            Inherits Syntax.InternalSyntax.VisualBasicSyntaxRewriter

            Public Overrides Function VisitIdentifierName(node As InternalSyntax.IdentifierNameSyntax) As InternalSyntax.VisualBasicSyntaxNode
                If node.Identifier.ValueText.StartsWith("a", StringComparison.Ordinal) Then
                    Return InternalSyntax.SyntaxFactory.IdentifierName(DirectCast(InternalSyntax.SyntaxFactory.Identifier("b" + node.Identifier.ValueText.Substring(1)).
                                                                                      WithLeadingTrivia(node.Identifier.GetLeadingTrivia()).
                                                                                      WithTrailingTrivia(node.Identifier.GetTrailingTrivia()), InternalSyntax.IdentifierTokenSyntax))
                End If

                Return MyBase.VisitIdentifierName(node)
            End Function
        End Class
    End Class
End Namespace
